### PR TITLE
Collect multiple errors on validation run

### PR DESCRIPTION
**Pull request by egor-denysenko**
## Description

The goal of this PR is to propose an approach to allow struct validation pass to group errors and return all of them at execution end.

Errors will be served as formatted strings separated by newlines. 
With the possibility from client side to parse errors one by one with a for loop and errors package.

Usage examples:

```go
// Existing code continues to work
if err := ValidateUser(user); err \!= nil {
    log.Printf("Validation failed: %v", err)
}

// Access detailed errors when needed
if err := ValidateUser(user); err \!= nil {
    var validationErrors govalid.ValidationErrors
    if errors.As(err, &validationErrors) {
        for _, e := range validationErrors {
            log.Printf("Field %s: %s", e.Path, e.Reason)
        }
    }
}
```

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update (TODO)
- [X] Refactoring (no functional changes)

## Quality Checklist

Before submitting this pull request, ensure your contribution meets these quality standards:

### Code Quality
- [X] Code passes `make golangci-lint` without errors
- [X] All tests pass: `make test`
- [X] Code follows existing patterns and conventions
- [X] No hard-coded values (use constants or configuration)
- [X] Error messages are clear and actionable

### Testing Requirements
- [X] Golden tests created and passing (`internal/analyzers/govalid/testdata/`)
- [X] Unit tests implemented with boundary value testing (`test/unit/`)
- [X] Benchmark tests comparing against popular validation libraries (`test/benchmark/`)
  - [X] [go-playground/validator](https://github.com/go-playground/validator) (BenchmarkGoPlayground*)
  - [ ] [asaskevich/govalidator](https://github.com/asaskevich/govalidator) (BenchmarkGoValidator*)
  - [ ] [go-ozzo/ozzo-validation](https://github.com/go-ozzo/ozzo-validation) (BenchmarkOzzoValidation*)
  - [ ] [gookit/validate](https://github.com/gookit/validate) (BenchmarkGookitValidate*)
- [X] All test files generated: `cd test && go generate`

### Performance Standards
- [ ] Zero allocations in validation logic (verified via benchmarks)
- [ ] Performance improvement over existing validation libraries (minimum 2x faster)
- [x] Benchmark results added to `test/benchmark/README.md`

### Documentation
- [x] README.md updated with new marker documentation
- [x] Code includes appropriate comments and examples
- [x] Validator usage examples provided

### Integration
- [ ] Validator scaffold generated with `make generate-validator MARKER=yourmarker`
- [ ] Registry files automatically updated
- [x] Binary rebuilt and installed: `go install ./cmd/govalid/`

### Final Verification
- [ ] All GitHub Actions CI checks pass
- [x] No breaking changes to existing functionality
- [x] Contribution follows project license requirements

## Additional Notes

PR aims to implement support for multiple errors aggregation when performing struct validation.

## Related Issues

Proposes implementation for #84

### CODE REVIEW

**Comment by sivchari** [comment]
Hi @ccoVeille @taua-almeida @BlackBuck
Please review it together. This PR will be turning point for us!!!

File: internal/validator/rules/alpha.go [hunk 1]
```diff
@@ -37,9 +37,18 @@ func (v *alphaValidator) Err() string {
...
+	const errTemplate = `
+		// [ERRVARIABLE] is the error returned when field @ is not alphabetic.
+		[ERRVARIABLE] = govaliderrors.ValidationError{Reason:"field @ must be alphabetic",Path:"PATH"}
+	`
+
+	replacer := strings.NewReplacer(
+		"[ERRVARIABLE]", v.ErrVariable(),
+		"@", v.FieldName(),
+		"PATH", fmt.Sprintf("%s.%s", v.structName, v.structName+v.FieldName()),
+	)
```
Comment (ccoVeille) [inline internal/validator/rules/alpha.go:49]: I feel like there are 3 pattern here. Only one should be used

```suggestion
	replacer := strings.NewReplacer(
		"[ERRVARIABLE]", v.ErrVariable(),
		"[FIELD]", v.FieldName(),
		"[PATH]", fmt.Sprintf("%s.%s", v.structName, v.structName+v.FieldName()),
	)
```

I report it because I had always found @ strange and now others are added, it feel like it should be fixed and consistent.

Also, I would prefer avoiding a pattern that could collide with field name, so I would prefer something that couldn't break later if description/comments are added and autogenerated. I'm afraid something like "this variable would be use in PATH" with the logic of Unix PATH variable, and it would be replaced

This format is more likely to never collide with something the user may provide

```suggestion
	replacer := strings.NewReplacer(
		"[@ERRVARIABLE]", v.ErrVariable(),
		"[@FIELD]", v.FieldName(),
		"[@PATH]", fmt.Sprintf("%s.%s", v.structName, v.structName+v.FieldName()),
	)
```

File: internal/validator/rules/cel.go [hunk 2]
```diff
@@ -50,27 +50,27 @@ func (c *celValidator) FieldName() string {
...
+	`
 
-	// Replace EXPRESSION placeholder with the actual CEL expression
-	errorString := result.String()
-	errorString = strings.ReplaceAll(errorString, "EXPRESSION", c.expression)
+	replacer := strings.NewReplacer(
+		"[ERRVARIABLE]", c.ErrVariable(),
+		"@", fieldName,
+		"PATH", fmt.Sprintf("%s.%s", c.structName, fieldName),
+		"EXPRESSION", c.expression,
```
Comment (ccoVeille) [inline internal/validator/rules/cel.go:70]: Same here and so on ...

File: internal/validator/rules/minitems.go [hunk 3]
```diff
@@ -34,15 +34,28 @@ func (m *minItemsValidator) FieldName() string {
 
 func (m *minItemsValidator) Err() string {
 	key := fmt.Sprintf(minItemsKey, m.structName+m.FieldName())
+
+	var result strings.Builder
+
 	if validator.GeneratorMemory[key] {
-		return ""
+		return result.String()
 	}
```
Comment (ccoVeille) [inline internal/validator/rules/minitems.go:43]: I don't understand the need of calling a stringer of a variable that will return empty string

This would be clearer I think

```go
        if validator.GeneratorMemory[key] { 
               return ""
        } 
 
        var result strings.Builder
```

Comment (egor-denysenko) [reply internal/validator/rules/minitems.go:43]: Nice catch, leftover from previous attempts before stabilising with `strings.Replacer` pattern.
Will remove in next pr's

File: internal/validator/rules/alpha.go (hunk 1 above)
Comment (egor-denysenko) [reply internal/validator/rules/alpha.go:49]: > I report it because I had always found @ strange and now others are added, it feel like it should be fixed and consistent.
Agree better maintain consistency with the @ that was already in there.
If after this one it doesn't look good it can be changed later on.

File: validation/errors/errors.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply validation/errors/errors.go:1]: Yep, missed this part in docs example https://sivchari.github.io/govalid/examples/#:~:text=Advanced%20Error%20Handling%20with%20errors.Is.
Will adapt new struct to implement also Is method from errors package.

Speaking of docs, also those should be updated, added to TODO's

File: internal/markers/markers_generated.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply internal/markers/markers_generated.go:1]: Interestingly those changes are made by tooling itself (govalid installed from repo) as soon as i launch make generate-validator.

Here is a dump of from clean main branch to assert that behaviour is the same as in current branch and previous merge created some instability and discrepancy in my branch.

```bash
govalid on  main via  v1.24.3
➜ make fmt && make generate-validator && make generate-validation-code
go tool -modfile tools/go.mod github.com/golangci/golangci-lint/v2/cmd/golangci-lint fmt ./...
go run cmd/generate-validators/main.go -marker=
✓ Generated initializers for 17 validators
go install ./cmd/govalid
package github.com/sivchari/govalid/cmd/govalid
        imports github.com/sivchari/govalid/internal/analyzers/govalid
        imports github.com/sivchari/govalid/internal/validator/registry/initializers: case-insensitive file name collision: "maxItems.go" and "maxitems.go"
make: *** [Makefile:36: install-govalid] Error 1
```
This also applies for the comments below https://github.com/sivchari/govalid/pull/121#discussion_r2246741186.

To solve this on main branch i had to:
```bash
govalid on  main [!?] via  v1.24.3
➜ rm internal/validator/registry/initializers/minItems.go internal/validator/registry/initializers/maxItems.go internal/validator/registry/initializers/minLength.go

govalid on  main [✘!?] via  v1.24.3
➜ make fmt && make generate-validator && make generate-validation-code
go tool -modfile tools/go.mod github.com/golangci/golangci-lint/v2/cmd/golangci-lint fmt ./...
go run cmd/generate-validators/main.go -marker=
✓ Generated initializers for 17 validators
go install ./cmd/govalid
go generate ./test/marker.go

govalid on  main [✘!?] via  v1.24.3 took 2s
➜ git status
On branch main
Your branch is up to date with 'origin/main'.

You are in a sparse checkout with 100% of tracked files present.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   internal/markers/markers_generated.go
        deleted:    internal/validator/registry/initializers/maxItems.go
        deleted:    internal/validator/registry/initializers/minItems.go
        deleted:    internal/validator/registry/initializers/minLength.go

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        internal/validator/registry/initializers/maxitems.go  internal/validator/registry/initializers/minitems.go  internal/validator/registry/initializers/minlength.go

no changes added to commit (use "git add" and/or "git commit -a")
```
After removing conflicting files status is same as branch.

Will investigate a if issue is still on my end 🤔 now i got curious on why it is happening

File: internal/analyzers/govalid/templates/validation.go.tmpl [hunk 4]
```diff
@@ -27,27 +28,40 @@ func Validate{{.TypeName}}(t *{{.TypeName}}) error {
...
 	    	{
   			t := t.{{ $parentVariable }}
 		{{ end -}}
+
 		{{ range .Validators }}
 			{{ if ne .Validate "" }}
 				if {{.Validate}} {
-			    	return {{ .ErrVariable }}
+  			  err := {{.ErrVariable}}
+					err.Type = "{{$.TypeName}}"
```
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply internal/analyzers/govalid/templates/validation.go.tmpl]: Solved in [d5b6cad](https://github.com/sivchari/govalid/pull/121/commits/d5b6cadd7f17a0c8761b4742559d6da6a962d574)

File: internal/validator/rules/alpha.go (hunk 1 above)
Comment (egor-denysenko) [reply internal/validator/rules/alpha.go]: [76c06d9](https://github.com/sivchari/govalid/pull/121/commits/76c06d91c0787734ab214f514301da2f5edc5792) solved with

File: internal/validator/rules/minitems.go (hunk 3 above)
Comment (egor-denysenko) [reply internal/validator/rules/minitems.go]: [76c06d9](https://github.com/sivchari/govalid/pull/121/commits/76c06d91c0787734ab214f514301da2f5edc5792) leftover removed in

File: internal/markers/markers_generated.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply internal/markers/markers_generated.go:1]: https://github.com/egor-denysenko/govalid/blob/main/cmd/generate-validators/internal/generate/discovery.go#L90 seems that this line is the one responsible for giving file names to validators initializers.
And probably generation wasn't done recently and error staid silent until now.

[the remark this answers is not in the document]
Comment (egor-denysenko) [reply internal/markers/markers_generated.go:1]: Ok did some tests on `main` to prove the point and it seems that
https://github.com/sivchari/govalid/blob/main/cmd/generate-validators/internal/generate/discovery.go#L90 is the cause for file namespace conflicts.

After removing that line files will be maintained in camel case as initially expected, and fix current discrepancy in initializers having some of those in camel case and some of those in lowercase.
https://github.com/sivchari/govalid/tree/main/internal/validator/registry/initializers

**Review by ccoVeille — APPROVED** [review]

File: validation/errors/errors.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply validation/errors/errors.go:1]: Both ValidationError and ValidationErrors Is method implementation was done in [85b0d72](https://github.com/sivchari/govalid/pull/121/commits/85b0d72446bfda4776db085f729068fcbb861f74).

[the remark this answers is not in the document]
Comment (egor-denysenko) [reply validation/errors/errors.go:1]: During implementation of `Is` method I've noticed that structs with multiple validation markers on an attribute are not generated properly.
This seems to affect main branch and consequently also this feature branches are not working properly since we are generating only the last validation rule from a list of struct attribute markers and not all.

steps to reproduce that:

Clone brand new repo
```bash

git clone https://github.com/sivchari/govalid.git

cd govalid

# Ensure fresh installation
which govalid | xargs rm

go install ./... 

# Feedback: in readme go install method could be also shown and not only in docs website
# go install github.com/sivchari/govalid/cmd/govalid@latest
```
create example module and file.
```bash
mkdir /tmp/regression && cd /tmp/regression && go mod init test && echo 'package main

//go:generate govalid .
type User struct {
        // +govalid:required
        Name string `json:"name"`

        // +govalid:required
        // +govalid:email
        Email string `json:"email"`

        // +govalid:gte=18
        Age int `json:"age"`
}

func main() {}' > main.go
```
Create validators with tooling
```bash
go generate ./...
```
Generated output 
```bash
/tmp/tmp-govalid via  v1.24.3
➜ cat main_user_validator.go
// Code generated by govalid; DO NOT EDIT.
package main

import (
        "errors"
        "github.com/sivchari/govalid/validation/validationhelper"
)

var (
        // ErrNilUser is returned when the User is nil.
        ErrNilUser = errors.New("input User is nil")

        // ErrUserNameRequiredValidation is returned when the UserName is required but not provided.
        ErrUserNameRequiredValidation = errors.New("field UserName is required")

        // ErrUserEmailEmailValidation is the error returned when the field is not a valid email address.
        ErrUserEmailEmailValidation = errors.New("field UserEmail must be a valid email address")

        // ErrUserAgeGTEValidation is the error returned when the value of the field is less than 18.
        ErrUserAgeGTEValidation = errors.New("field UserAge must be greater than or equal to 18")
)

func ValidateUser(t *User) error {
        if t == nil {
                return ErrNilUser
        }

        if t.Name == "" {
                return ErrUserNameRequiredValidation
        }

        if !validationhelper.IsValidEmail(t.Email) {
                return ErrUserEmailEmailValidation
        }

        if !(t.Age >= 18) {
                return ErrUserAgeGTEValidation
        }

        return nil
}
```
The output should as stated per docs for Email struct attribute have two validators and consequently two Errors generated. 
But as it seems only last one is generated 
```go
        // +govalid:required
        // +govalid:email
        Email string `json:"email"`
```
If we were to excange the order of the fields only required would be generated.

```bash
rm ./* && go mod init test && echo 'package main

//go:generate govalid .
type User struct {
        // +govalid:required
        Name string `json:"name"`

        // +govalid:email
        // +govalid:required
        Email string `json:"email"`

        // +govalid:gte=18
        Age int `json:"age"`
}

func main() {}' >> main.go
```
and then regenerate validators
```bash
go generate ./...
```
Confirms that only last marker is taken in consideration
```bash
/tmp/tmp-govalid via  v1.24.3
➜ cat main_user_validator.go
// Code generated by govalid; DO NOT EDIT.
package main

import (
        "errors"
)

var (
        // ErrNilUser is returned when the User is nil.
        ErrNilUser = errors.New("input User is nil")

        // ErrUserNameRequiredValidation is returned when the UserName is required but not provided.
        ErrUserNameRequiredValidation = errors.New("field UserName is required")

        // ErrUserEmailRequiredValidation is returned when the UserEmail is required but not provided.
        ErrUserEmailRequiredValidation = errors.New("field UserEmail is required")

        // ErrUserAgeGTEValidation is the error returned when the value of the field is less than 18.
        ErrUserAgeGTEValidation = errors.New("field UserAge must be greater than or equal to 18")
)

func ValidateUser(t *User) error {
        if t == nil {
                return ErrNilUser
        }

        if t.Name == "" {
                return ErrUserNameRequiredValidation
        }

        if t.Email == "" {
                return ErrUserEmailRequiredValidation
        }

        if !(t.Age >= 18) {
                return ErrUserAgeGTEValidation
        }

        return nil
}
```

---

Speaking of which since it's probably a regression introduced in some previous commit is shouldn't block the review process of this PR. 

But an issue could be raised, fixed and shipped together with this feature in new minor release.

Or even better fixing this regression, release patch, and then provide this feature again. 

I would go with the latter but probably it should be discussed to check out more cons and pros about it.

File: internal/analyzers/govalid/testdata/src/lte/govalid.golden [hunk 5]
```diff
@@ -3,43 +3,58 @@ package lte
...
+	govaliderrors "github.com/sivchari/govalid/validation/errors"
 )
 
 var (
 	// ErrNilLTE is returned when the LTE is nil.
 	ErrNilLTE = errors.New("input LTE is nil")
 
 	// ErrLTEAgeLTEValidation is the error returned when the value of the field is greater than 100.
-	ErrLTEAgeLTEValidation = errors.New("field LTEAge must be less than or equal to 100")
+	ErrLTEAgeLTEValidation = govaliderrors.ValidationError{Reason: "field Age must be less than or equal to 100", Path: "LTE.Age"}
```
[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden]: This is because one of them is the struct name and the other is the name of the marker... it causes confusion when the struct name used is the same as the marker name.

[the remark this answers is not in the document]
Comment (ccoVeille) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden]: I assumed it, bit can't we avoid this ?

Maybe it's a niche issue anyway

[the remark this answers is not in the document]
Comment (BlackBuck) [reply internal/analyzers/govalid/testdata/src/lte/govalid.golden]: Yeah. I confirmed this earlier and, apparently, [it's expected behavior](https://github.com/sivchari/govalid/pull/79#issuecomment-3125516135)

**Review by taua-almeida — APPROVED** [review]
loved it

File: validation/errors/errors.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply validation/errors/errors.go:1]: Quick update since i was mentioning my previous TODO https://github.com/sivchari/govalid/pull/121/#discussion_r2248971428
Also docs were updated in [661a4ee](https://github.com/sivchari/govalid/pull/121/commits/661a4eeb7619874109cfbe6491229c9d03cb6ce6)

File: internal/markers/markers_generated.go
[the remark this answers is not in the document]
Comment (egor-denysenko) [reply internal/markers/markers_generated.go:1]: @sivchari How is the preferred method for handling this discrepancy? 
Tried with latest main as of writing [0693ab4](https://github.com/sivchari/govalid/commit/0693ab46c4e2afb0060f01fd23b8199e71c44aa0) and namespace conflict is still in there.

Proposed solution:
Use this PR to stabilize markers generator with all lower case initializers file names.

And follow up with another issue and consequent PR to restore wanted behaivor of having camel case files. 

Mostly to not expand even more contest of this way too big PR of mine 🙇🏻

**Comment by egor-denysenko** [comment]
Ok seems that PR has come closer to a final rewiable state.

Will mark it as ready for review.

Feel free to add any additional comments on anything that it's not quite clear or consistent 🙇🏻

File: internal/markers/markers_generated.go
[the remark this answers is not in the document]
Comment (sivchari) [reply internal/markers/markers_generated.go:1]: Okay, I totally understand it. This is gap between OSs. I use macOS, then it doesn't distinguish whether these literals are upper-case or lower-case.

**Comment by sivchari** [comment]
@egor-denysenko 
Please rebase this pr finally, just in case . Unrelated changes would be removed.

**Review by ccoVeille — APPROVED** [review]

**Review by sivchari — APPROVED** [review]
Hey @egor-denysenko , this is kudos for govalid and Go community !!!!
You completed indispensable feature. Thank you for your efforts, really really appreciated!
